### ISSUE TITLE
Getting Permission Denied when deregistering service with management token

### ISSUE THREAD

**Issue by beardedeagle** [hashicorp/consul#3157]
### `consul version` for both Client and Server
Server: `0.8.4`

### `consul info` for Server
Server:
```
agent:
	check_monitors = 1
	check_ttls = 0
	checks = 1
	services = 2
build:
	prerelease =
	revision = f436077
	version = 0.8.4
consul:
	bootstrap = false
	known_datacenters = 2
	leader = true
	leader_addr = REDACTED:8300
	server = true
raft:
	applied_index = 4109
	commit_index = 4109
	fsm_pending = 0
	last_contact = 0
	last_log_index = 4109
	last_log_term = 11
	last_snapshot_index = 0
	last_snapshot_term = 0
	latest_configuration = [{Suffrage:Voter ID:feb33bf5-61a7-2575-f09a-b1454235ef29 Address:REDACTED:8300} {Suffrage:Voter ID:e84e3941-f0cf-79cf-2661-dfc89c41f2ee Address:REDACTED:8300} {Suffrage:Voter ID:1bdfaf31-ffa1-90a2-46c8-2ed18abbd332 Address:REDACTED:8300} {Suffrage:Voter ID:6a4ddb06-4547-aeb3-33b8-b072398d7ab0 Address:REDACTED:8300} {Suffrage:Voter ID:38e17d51-02a7-86ef-28a2-48f403dde214 Address:REDACTED:8300}]
	latest_configuration_index = 1
	num_peers = 4
	protocol_version = 3
	protocol_version_max = 3
	protocol_version_min = 0
	snapshot_version_max = 1
	snapshot_version_min = 0
	state = Leader
	term = 11
runtime:
	arch = amd64
	cpu_count = 4
	goroutines = 185
	max_procs = 4
	os = linux
	version = go1.8.3
serf_lan:
	coordinate_resets = 0
	encrypted = true
	event_queue = 0
	event_time = 3
	failed = 0
	health_score = 0
	intent_queue = 0
	left = 0
	member_time = 45
	members = 9
	query_queue = 0
	query_time = 1
serf_wan:
	coordinate_resets = 0
	encrypted = true
	event_queue = 0
	event_time = 1
	failed = 0
	health_score = 0
	intent_queue = 0
	left = 0
	member_time = 39
	members = 10
	query_queue = 0
	query_time = 1
```

### Operating system and Environment details
```
/etc/redhat-release
CentOS Linux release 7.3.1611 (Core)

uname -a
Linux p3dlconsul01.cloud.phx3.gdg 3.10.0-514.21.1.el7.x86_64 #1 SMP Thu May 25 17:04:51 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux

client package:
python-consul==0.7.0
```

### Server config
```
{
	"acl_agent_token": "ACL_MASTER_TOKEN",
	"acl_datacenter": "p3",
	"acl_default_policy": "deny",
	"acl_down_policy": "extend-cache",
	"acl_master_token": "ACL_MASTER_TOKEN",
	"advertise_addr": "1.2.3.4",
	"bind_addr": "1.2.3.4",
	"bootstrap_expect": 5,
	"client_addr": "0.0.0.0",
	"data_dir": "/var/lib/consul",
	"datacenter": "p3",
	"disable_update_check": true,
	"enable_syslog": true,
	"encrypt": "ENCRYPT_TOKEN",
	"log_level": "INFO",
	"node_meta": {
		"datacenter": "p3",
		"environment": "dev",
		"type": "server"
	},
	"node_name": "node1",
	"performance": {
		"raft_multiplier": 1
	},
	"protocol": 3,
	"raft_protocol": 3,
	"rejoin_after_leave": true,
	"retry_interval": "30s",
	"retry_join": ["node1.cloud.phx3.gdg", "node2.cloud.phx3.gdg", "node3.cloud.phx3.gdg", "node4.cloud.phx3.gdg", "node5.cloud.phx3.gdg"],
	"server": true,
	"start_join": ["node1.cloud.phx3.gdg", "node2.cloud.phx3.gdg", "node3.cloud.phx3.gdg", "node4.cloud.phx3.gdg", "node5.cloud.phx3.gdg"],
	"ui": true
}
```

### Description of the Issue (and unexpected/desired result)
I have v8 acls enabled with a acl policy of deny. I have a management acl with no explicit rules set that I am able to use to register services with healtchecks. However when I try to deregister the services I get a `permission denied`. If I add the following to the anonymous acl `service "HAC" {policy = "write"}` my management acl will be able to successfully deregister the service. I also tried using the master acl token, with the same results.

EDIT: Registering and deregistering is happening via calls to the agent api with `python-consul`, just to clarify.

### Reproduction steps
Set up consul quorum with v8 acls enabled and a policy of deny. Create a management token with no explicit rules. Register a service. Try to deregister the service with that management token.

### Log Fragments
```
Jun 16 14:37:02 REDACTED consul[3638]: http: Request GET /v1/agent/service/deregister/HAC001D_p3_REDACTED, error: Permission denied from=REDACTED:56005
```

**Comment by preetapan** [comment]
@beardedeagle this looks like its expected behavior - according to https://github.com/hashicorp/consul/issues/1383#issuecomment-279733386 

Also see https://www.consul.io/docs/guides/acl.html#complete-acl-coverage-in-consul-0-8, particularly this 

_Consul 0.8 added many more ACL policy types and brought ACL enforcement to Consul agents for the first time. To ease the transition to Consul 0.8 for existing ACL users, there's a configuration option to disable these new features. To disable support for these new ACLs, set the acl_enforce_version_8 configuration option to false on Consul clients and servers._

**Comment by beardedeagle** [comment]
@preetapan Not sure #1383 addresses this. I _AM_ passing a token to the deregister call and that token as management access. Even if I create a token with explicit write access to the nodes I am trying to deregister, it will not work without applying it to the anonymous token.

**Comment by beardedeagle** [comment]
ok I _THINK_ I have this solved, though I am not sure I like the solution:
```
rules = """key "services/{}" {{policy = "write"}}
key "build/{}" {{policy = "write"}}
key "build/{}" {{policy = "write"}}
key "build/{}" {{policy = "write"}}
key "build/{}" {{policy = "write"}}
key "build/{}" {{policy = "write"}}
key "build/{}" {{policy = "write"}}
key "build/{}" {{policy = "write"}}
key "build/{}" {{policy = "write"}}
agent "{}" {{policy = "write"}}
agent "{}" {{policy = "write"}}
agent "{}" {{policy = "write"}}
agent "{}" {{policy = "write"}}
agent "{}" {{policy = "write"}}
agent "{}" {{policy = "write"}}
agent "{}" {{policy = "write"}}
agent "{}" {{policy = "write"}}
session "{}" {{policy = "write"}}
session "{}" {{policy = "write"}}
session "{}" {{policy = "write"}}
session "{}" {{policy = "write"}}
session "{}" {{policy = "write"}}
session "{}" {{policy = "write"}}
session "{}" {{policy = "write"}}
session "{}" {{policy = "write"}}
node "{}" {{policy = "write"}}
node "{}" {{policy = "write"}}
node "{}" {{policy = "write"}}
node "{}" {{policy = "write"}}
node "{}" {{policy = "write"}}
node "{}" {{policy = "write"}}
node "{}" {{policy = "write"}}
node "{}" {{policy = "write"}}
service "{}" {{policy = "write"}}
service "{}" {{policy = "write"}}
service "{}" {{policy = "write"}}
service "{}" {{policy = "write"}}
service "{}" {{policy = "write"}}
service "{}" {{policy = "write"}}
service "{}" {{policy = "write"}}
service "{}" {{policy = "write"}}
service "consul" {{policy = "write"}}""".format(jbod['service'],
                                                *[x['name'] for x in jbod['build']],
                                                *[x['name'] for x in jbod['build']],
                                                *[x['name'] for x in jbod['build']],
                                                *[x['name'] for x in jbod['build']],
                                                *['{}_{}_{}'.format(
                                                    x['group'], x['dc'], x['float']) for x in jbod['build']])

jbod['btoken'] = MASTER.acl.create(name=jbod['service'], rules=rules)

for idx in enumerate(jbod['build']):
    MASTER.kv.put('build/{}/token'.format(jbod['build'][idx[0]]['name']), jbod['btoken'])
```
I create a token for each "service" a user requests and assign it these acls (service means something different in my world). Then I use this token when I deregister the services and nodes.

**Comment by preetapan** [comment]
@beardedeagle Spent some time digging into this and not able to reproduce what you are seeing, I could be missing something so I've created a gist of all my config/payload used to register/deregister [here](https://gist.github.com/preetapan/e26183640acc5340e618ea1429e51176). Also https://gist.github.com/preetapan/fa7590e10a041e538c8669f70cdd97fa shows what I did to register/deregister. I didn't see any errors with the configs as described.
